File: bot.php
<?php

// ====== НАСТРОЙКИ ======
$botToken = "8597712382:AAE4-gQXxx22AE71-yAFJ4wLQ3cG5PiNmXA";
$staffGroupId = -1003280448019;
$allowedStaff = [1062756366];

$apiUrl = "https://api.telegram.org/bot$botToken/";

// ====== ФАЙЛЫ БД ======
$dbFile = __DIR__ . "/messages.json";
$topicsFile = __DIR__ . "/topics.json";

if (!file_exists($dbFile)) file_put_contents($dbFile, "{}");
if (!file_exists($topicsFile)) file_put_contents($topicsFile, "{}");

// ====== ФУНКЦИЯ API ======
function tgRequest($method, $params = []) {
    global $apiUrl;
    $ch = curl_init($apiUrl . $method);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($ch, CURLOPT_POSTFIELDS, $params);
    $response = curl_exec($ch);
    curl_close($ch);
    return json_decode($response, true);
}

// ====== РАБОТА С ТОПИКАМИ ======
function getThreadForClient($chatId, $groupTitle) {
    global $staffGroupId, $topicsFile;
    $topics = json_decode(file_get_contents($topicsFile), true);

    if (isset($topics["c_$chatId"])) return $topics["c_$chatId"];

    $topicName = "📂 " . $groupTitle;
    $res = tgRequest("createForumTopic", ["chat_id" => $staffGroupId, "name" => $topicName]);
    
    if (isset($res["result"]["message_thread_id"])) {
        $threadId = $res["result"]["message_thread_id"];
        $topics["c_$chatId"] = $threadId;
        $topics["t_$threadId"] = $chatId;
        file_put_contents($topicsFile, json_encode($topics, JSON_PRETTY_PRINT));
        return $threadId;
    }
    return null;
}

function getClientIdByThread($threadId) {
    global $topicsFile;
    $topics = json_decode(file_get_contents($topicsFile), true);
    return $topics["t_$threadId"] ?? null;
}

// ====== ПОИСК СООТВЕТСТВИЯ REPLY ======
function getStaffMsgIdByClientMsgId($clientMsgId) {
    global $dbFile;
    $db = json_decode(file_get_contents($dbFile), true);
    // Ищем в базе сообщение, которое было создано в группе сотрудников для этого сообщения клиента
    foreach ($db as $staffId => $data) {
        if ($data['client_message_id'] == $clientMsgId) {
            return $staffId;
        }
    }
    return null;
}

// ====== ПОЛУЧЕНИЕ ОБНОВЛЕНИЯ ======
$update = json_decode(file_get_contents("php://input"), true);
if (!$update || !isset($update["message"])) exit;

$msg = $update["message"];
$chatId = $msg["chat"]["id"];
$msgId  = $msg["message_id"];
$userId = $msg["from"]["id"];

$groupTitle = $msg["chat"]["title"] ?? "Личное (".$msg["from"]["first_name"].")";
$firstName = $msg["from"]["first_name"] ?? "";
$lastName = $msg["from"]["last_name"] ?? "";
$senderName = trim($firstName . " " . $lastName);
if (empty($senderName)) $senderName = "User_$userId";

// ====================
// ВХОДЯЩЕЕ ОТ СОТРУДНИКА
// ====================
if ($chatId == $staffGroupId) {
    if (!in_array($userId, $allowedStaff)) exit;

    $targetClientId = null;
    $replyToClientMsgId = null;
    $currentThreadId = $msg["message_thread_id"] ?? null;

    if (isset($msg["reply_to_message"])) {
        $db = json_decode(file_get_contents($dbFile), true);
        $map = $db[$msg["reply_to_message"]["message_id"]] ?? null;
        if ($map) {
            $targetClientId = $map["client_chat_id"];
            $replyToClientMsgId = $map["client_message_id"];
        }
    }
    
    if (!$targetClientId && $currentThreadId) {
        $targetClientId = getClientIdByThread($currentThreadId);
    }

    if ($targetClientId) {
        $method = "sendMessage";
        $params = ["chat_id" => $targetClientId];
        if ($replyToClientMsgId) $params["reply_to_message_id"] = $replyToClientMsgId;

        if (isset($msg["text"])) {
            $params["text"] = "👨‍💼 *Поддержка:*\n\n" . $msg["text"];
            $params["parse_mode"] = "Markdown";
        } elseif (isset($msg["photo"])) {
            $method = "sendPhoto";
            $params["photo"] = end($msg["photo"])["file_id"];
            $params["caption"] = "👨‍💼 Ответ поддержки";
        } elseif (isset($msg["video"])) {
            $method = "sendVideo";
            $params["video"] = $msg["video"]["file_id"];
        } elseif (isset($msg["document"])) {
            $method = "sendDocument";
            $params["document"] = $msg["document"]["file_id"];
        } elseif (isset($msg["voice"])) {
            $method = "sendVoice";
            $params["voice"] = $msg["voice"]["file_id"];
        }

        $result = tgRequest($method, $params);
        if (!$result || (isset($result["ok"]) && !$result["ok"])) {
            tgRequest("sendMessage", ["chat_id" => $staffGroupId, "message_thread_id" => $currentThreadId, "text" => "⚠️ Ошибка доставки"]);
        }
    }
    exit;
}

// ====================
// ВХОДЯЩЕЕ ОТ КЛИЕНТА
// ====================
$threadId = getThreadForClient($chatId, $groupTitle);

if ($threadId) {
    $method = "sendMessage";
    $params = [
        "chat_id" => $staffGroupId,
        "message_thread_id" => $threadId,
        "parse_mode" => "Markdown"
    ];

    // ПРОВЕРКА НА REPLY СО СТОРОНЫ КЛИЕНТА
    if (isset($msg["reply_to_message"])) {
        $originalClientMsgId = $msg["reply_to_message"]["message_id"];
        $correspondingStaffMsgId = getStaffMsgIdByClientMsgId($originalClientMsgId);
        if ($correspondingStaffMsgId) {
            $params["reply_to_message_id"] = $correspondingStaffMsgId;
        }
    }

    $prefix = "👤 *{$senderName}:*\n";

    if (isset($msg["text"])) {
        $params["text"] = $prefix . $msg["text"];
    } elseif (isset($msg["photo"])) {
        $method = "sendPhoto";
        $params["photo"] = end($msg["photo"])["file_id"];
        $params["caption"] = $prefix;
    } elseif (isset($msg["video"])) {
        $method = "sendVideo";
        $params["video"] = $msg["video"]["file_id"];
        $params["caption"] = $prefix;
    } elseif (isset($msg["document"])) {
        $method = "sendDocument";
        $params["document"] = $msg["document"]["file_id"];
        $params["caption"] = $prefix;
    } else {
        $copyParams = [
            "chat_id" => $staffGroupId,
            "from_chat_id" => $chatId,
            "message_id" => $msgId,
            "message_thread_id" => $threadId
        ];
        if (isset($params["reply_to_message_id"])) $copyParams["reply_to_message_id"] = $params["reply_to_message_id"];
        $sent = tgRequest("copyMessage", $copyParams);
        $resId = $sent["result"]["message_id"] ?? null;
    }

    if ($method !== "copyMessage" && !isset($resId)) {
        $sent = tgRequest($method, $params);
        $resId = $sent["result"]["message_id"] ?? null;
    }

    if ($resId) {
        $db = json_decode(file_get_contents($dbFile), true);
        $db[$resId] = [
            "client_chat_id" => $chatId,
            "client_message_id" => $msgId
        ];
        if (count($db) > 3000) $db = array_slice($db, -3000, null, true);
        file_put_contents($dbFile, json_encode($db, JSON_PRETTY_PRINT));
    }
}

File: bot.py
import logging
import json
import os
from aiogram import Bot, Dispatcher, executor, types

# --- Конфиг ---
BOT_TOKEN = "8597712382:AAE4-gQXxx22AE71-yAFJ4wLQ3cG5PiNmXA"
STAFF_GROUP_ID = -1003280448019
ALLOWED_STAFF = [1062756366]

# --- Логирование в файл и консоль ---
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(name)s: %(message)s",
    handlers=[
        logging.FileHandler("log.txt", encoding="utf-8"),
        logging.StreamHandler()
    ]
)
logger = logging.getLogger("bridge-bot")

bot = Bot(token=BOT_TOKEN)
dp = Dispatcher(bot)

# --- JSON база ---
DB_FILE = "messages.json"
TOPICS_FILE = "topics.json"  # соответствие клиентской группы -> message_thread_id

def load_json(path, default):
    if not os.path.exists(path):
        with open(path, "w", encoding="utf-8") as f:
            json.dump(default, f, ensure_ascii=False, indent=2)
        return default
    with open(path, "r", encoding="utf-8") as f:
        try:
            return json.load(f)
        except Exception:
            # если файл повреждён — создаём заново
            logging.error(f"Файл {path} повреждён, пересоздаю")
            with open(path, "w", encoding="utf-8") as w:
                json.dump(default, w, ensure_ascii=False, indent=2)
            return default

def save_json(path, data):
    with open(path, "w", encoding="utf-8") as f:
        json.dump(data, f, ensure_ascii=False, indent=2)

def get_client_key(chat: types.Chat) -> str:
    # Ключ для идентификации клиента: используем title если есть, иначе ID
    return chat.title or str(chat.id)

async def get_or_create_topic(client_key: str) -> int:
    topics = load_json(TOPICS_FILE, {})  # {client_key: topic_id}
    topic_id = topics.get(client_key)
    if topic_id:
        return topic_id

    # Создаём новую тему и сохраняем её ID
    try:
        new_topic = await bot.create_forum_topic(STAFF_GROUP_ID, name=client_key)
        topic_id = new_topic.message_thread_id
        topics[client_key] = topic_id
        save_json(TOPICS_FILE, topics)
        logger.info(f"Создана новая тема '{client_key}' с ID {topic_id}")
        return topic_id
    except Exception as e:
        logger.exception(f"Не удалось создать тему для '{client_key}': {e}")
        raise

@dp.message_handler(content_types=types.ContentTypes.ANY)
async def from_client(message: types.Message):
    # Игнорируем сообщения из группы сотрудников
    if message.chat.id == STAFF_GROUP_ID:
        return

    # Только групповые чаты клиентов (бот должен быть админом там)
    try:
        client_key = get_client_key(message.chat)
        topic_id = await get_or_create_topic(client_key)

        # Текст предварительного сообщения с метаданными клиента
        header = (
            f"📥 [Группа: {client_key}]\n"
            f"Клиент: {message.from_user.full_name}\n"
        )

        # Отправляем в тему сотрудников заголовок и само содержание (тип-сейф)
        await bot.send_message(
            STAFF_GROUP_ID,
            header,
            message_thread_id=topic_id
        )

        # Далее отправляем реальный контент клиента
        if message.text:
            sent = await bot.send_message(
                STAFF_GROUP_ID,
                message.text,
                message_thread_id=topic_id,
                reply_to_message_id=None
            )
        elif message.photo:
            sent = await bot.send_photo(
                STAFF_GROUP_ID,
                message.photo[-1].file_id,
                caption=message.caption or "",
                message_thread_id=topic_id
            )
        elif message.document:
            sent = await bot.send_document(
                STAFF_GROUP_ID,
                message.document.file_id,
                caption=message.caption or "",
                message_thread_id=topic_id
            )
        elif message.voice:
            sent = await bot.send_voice(
                STAFF_GROUP_ID,
                message.voice.file_id,
                caption=message.caption or "",
                message_thread_id=topic_id
            )
        elif message.video:
            sent = await bot.send_video(
                STAFF_GROUP_ID,
                message.video.file_id,
                caption=message.caption or "",
                message_thread_id=topic_id
            )
        elif message.sticker:
            sent = await bot.send_sticker(
                STAFF_GROUP_ID,
                message.sticker.file_id,
                message_thread_id=topic_id
            )
        else:
            sent = await bot.send_message(
                STAFF_GROUP_ID,
                "⚠️ Тип сообщения клиента не поддержан ботом.",
                message_thread_id=topic_id
            )

        # Сохраняем соответствие: staff_msg_id -> client_chat_id / client_msg_id / topic_id
        db = load_json(DB_FILE, {})
        db[str(sent.message_id)] = {
            "client_chat_id": message.chat.id,
            "client_msg_id": message.message_id,
            "topic_id": topic_id
        }
        save_json(DB_FILE, db)
        logger.info(f"Клиентское сообщение {message.message_id} → staff_msg {sent.message_id} в теме {topic_id}")
    except Exception:
        logger.exception("Ошибка при обработке сообщения клиента")

@dp.message_handler(lambda m: m.chat.id == STAFF_GROUP_ID, content_types=types.ContentTypes.ANY)
async def from_staff(message: types.Message):
    # Только разрешённые сотрудники
    if message.from_user.id not in ALLOWED_STAFF:
        logger.warning(f"Пользователь {message.from_user.id} пытался ответить без прав")
        return

    # Ответ должен быть реплаем на клиентский контент в теме
    if not message.reply_to_message:
        return

    try:
        db = load_json(DB_FILE, {})
        mapping = db.get(str(message.reply_to_message.message_id))
        if not mapping:
            # Нечему сопоставить — возможно, ответили на заголовок или старое сообщение
            logger.info("Не найдено соответствие для reply_to_message_id, пропускаю")
            return

        client_chat_id = mapping["client_chat_id"]

        # Отправляем ответ сотрудника в клиентскую группу
        if message.text:
            await bot.send_message(
                client_chat_id,
                f"👨‍💼 От сотрудника {message.from_user.full_name}: {message.text}"
            )
        elif message.photo:
            await bot.send_photo(
                client_chat_id,
                message.photo[-1].file_id,
                caption=f"👨‍💼 От сотрудника {message.from_user.full_name}" + (f"\n{message.caption}" if message.caption else "")
            )
        elif message.document:
            await bot.send_document(
                client_chat_id,
                message.document.file_id,
                caption=f"👨‍💼 От сотрудника {message.from_user.full_name}" + (f"\n{message.caption}" if message.caption else "")
            )
        elif message.voice:
            await bot.send_voice(
                client_chat_id,
                message.voice.file_id,
                caption=f"👨‍💼 От сотрудника {message.from_user.full_name}"
            )
        elif message.video:
            await bot.send_video(
                client_chat_id,
                message.video.file_id,
                caption=f"👨‍💼 От сотрудника {message.from_user.full_name}" + (f"\n{message.caption}" if message.caption else "")
            )
        elif message.sticker:
            await bot.send_sticker(
                client_chat_id,
                message.sticker.file_id
            )
        else:
            await bot.send_message(
                client_chat_id,
                f"👨‍💼 От сотрудника {message.from_user.full_name}: [тип сообщения не поддержан]"
            )

        logger.info(f"Ответ сотрудника {message.from_user.id} доставлен в {client_chat_id}")
    except Exception:
        logger.exception("Ошибка при обработке ответа сотрудника")

if __name__ == "__main__":
    logger.info("Бот запускается")
    executor.start_polling(dp, skip_updates=True)
